pkg/streamingpromql/evaluator.go

```diff
@@ -0,0 +1,256 @@
+	// chunks from store-gateways or ingesters.
+	ctx = limiter.AddMemoryTrackerToContext(ctx, e.memoryConsumptionTracker)
+
+	ctx, cancel := context.WithCancelCause(ctx)
+	e.cancel = cancel
+
+	if e.engine.timeout != 0 {
+		var cancelTimeoutCtx context.CancelFunc
+		ctx, cancelTimeoutCtx = context.WithTimeoutCause(ctx, e.engine.timeout, fmt.Errorf("%w: query timed out", context.DeadlineExceeded))
+
+		defer cancelTimeoutCtx()
+	}
+
+	// The order of the deferred cancellations is important: we want to cancel with errQueryFinished first, so we must defer this cancellation last
+	// (so that it runs before the cancellation of the context with timeout created above).
+	defer cancel(errQueryFinished)
+
+	queryID, err := e.engine.activeQueryTracker.Insert(ctx, e.activityTrackerDescription)
+	if err != nil {
+		return err
+	}
+
+	defer e.engine.activeQueryTracker.Delete(queryID)
+
+	err = e.root.Prepare(ctx, &types.PrepareParams{QueryStats: e.stats})
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	switch root := e.root.(type) {
+	case types.InstantVectorOperator:
+		if err := e.evaluateInstantVectorOperator(ctx, root, observer); err != nil {
+			return err
+		}
+
+	case types.RangeVectorOperator:
+		if err := e.evaluateRangeVectorOperator(ctx, root, observer); err != nil {
+			return err
+		}
+
+	case types.ScalarOperator:
+		if err := e.evaluateScalarOperator(ctx, root, observer); err != nil {
+			return err
+		}
+
+	case types.StringOperator:
+		if err := e.evaluateStringOperator(root, observer); err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("operator type %T produces unknown result type", root)
+	}
+
+	// To make comparing to Prometheus' engine easier, only return the annotations if there are some, otherwise, return nil.
+	if len(*e.annotations) == 0 {
+		e.annotations = nil
+	}
+
+	if err := observer.EvaluationCompleted(e, e.annotations, e.stats); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *Evaluator) evaluateInstantVectorOperator(ctx context.Context, o types.InstantVectorOperator, observer EvaluationObserver) error {
+	series, err := o.SeriesMetadata(ctx)
+	if err != nil {
+		return err
+	}
+
+	seriesCount := len(series) // Take the length now, as the observer takes ownership of the slice when SeriesMetadataEvaluated is called.
+
+	if err := observer.SeriesMetadataEvaluated(e, series); err != nil {
+		return err
+	}
+
+	for seriesIdx := range seriesCount {
+		d, err := o.NextSeries(ctx)
+		if err != nil {
+			if errors.Is(err, types.EOS) {
+				return fmt.Errorf("expected %v series, but only received %v", seriesCount, seriesIdx)
+			}
+
+			return err
+		}
+
+		if err := observer.InstantVectorSeriesDataEvaluated(e, seriesIdx, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *Evaluator) evaluateRangeVectorOperator(ctx context.Context, o types.RangeVectorOperator, observer EvaluationObserver) error {
```

Comment:
nit: since we have an operator type and observer type, WDYT about using `op` instead of `o` for operators so it's more obvious what it is?